daemons/controld/controld_remote_ra.c

```diff
@@ -630,13 +718,42 @@ remote_lrm_op_callback(lrmd_event_data_t * op)
             } else {
                 crm_trace("can't reschedule start, remaining timeout too small %d",
                           cmd->remaining_timeout);
+                purge_remote_rsc(op->remote_nodename, NULL);
                 pcmk__format_result(&(cmd->result), PCMK_OCF_UNKNOWN_ERROR,
                                     PCMK_EXEC_TIMEOUT,
                                     "%s without enough time to retry",
                                     pcmk_strerror(op->connection_rc));
             }
 
         } else {
+            lrm_state_t *connection_rsc = lrm_state_find(op->remote_nodename);
+
+            if (connection_rsc && connection_rsc->remote_ra_data) {
+                remote_ra_data_t *ra_data = connection_rsc->remote_ra_data;
+                time_t now = time(NULL);
+
+                /* The remote resource came back.  We want to mark its resource as
+                 * unexpired.  This is the first step in preserving its transient
+                 * attributes.  The second step happens in remote_node_up, which will
+                 * be called after this.
+                 *
+                 * However, the following conditions must be met:
+                 *
+                 * (1) We were waiting on seeing a start action for this resource, in
+                 * which case it is marked as expired already.
+                 *
+                 * (2) The uptime we got from the remote server's handshake indicates
+                 * it has been running longer than our timeout.  This implies that the
+                 * remote was running long enough to where it's probably simply been
+                 * unreachable and did not restart.
+                 */
+                if (ra_data->when_stopped <= now - EXPIRATION_TIMEOUT) {
```

Comment:
We definitely don't want to assume the node didn't restart based on a fixed time.

When the original connection host goes away, every surviving node will detect that via the peer callback. I'm thinking every node should memorize the timestamp at this point (likely a new field in crm_peer_t), since any node could be chosen as the new connection host. This is the time that we'll need to know the remote node was up longer than.

The controllers won't have any way of associating that dead node with the remote connection, so the scheduler will have to pass that information in the transition graph (the pseudo-stop is a good choice).

The DC (possibly newly elected, if the original DC was the host that died) will invoke the scheduler and get the transition graph back. The DC can check for any pseudo-stops with the new meta-attributes, and whether there is a start of the remote connection scheduled in the same transition. If no to either, we can let the purge happen normally.

If there is a start scheduled, then when the DC sends that command to the new connection host (which will likely not be the same as where the connection failed, because that node died and is probably not back up yet), it should add enough info for the new connection host to handle it from here. That means we're OK if the new DC dies before the process completes.

The new connection host will receive and initiate the start command. If the start completes successfully and the remote node was up since the drop, we can leave the attributes alone, otherwise we can purge them.

There's probably more corner cases to think about, but I think that should be the basic idea.

(BTW, it occurs to me that we should keep attributes alive only for remote nodes, not guest nodes or bundle nodes. I'm guessing that will happen automatically since they should hit XPATH_PSEUDO_FENCE.)

Reply:
> When the original connection host goes away, every surviving node will detect that via the peer callback. I'm thinking every node should memorize the timestamp at this point (likely a new field in crm_peer_t), since any node could be chosen as the new connection host. This is the time that we'll need to know the remote node was up longer than.

I assume this is going to have to be a GList or GHashTable, since there could potentially be several remotes that have problems.

> The DC (possibly newly elected, if the original DC was the host that died) will invoke the scheduler and get the transition graph back. The DC can check for any pseudo-stops with the new meta-attributes, and whether there is a start of the remote connection scheduled in the same transition. If no to either, we can let the purge happen normally.

I didn't see a great way of checking for a start action in the same transition from down in remote_ra_process_pseudo.  So what I did there was add the remote_expired marker and then later on handle actually purging the attributes.  The new block in remote_lrm_op_callback is meant to handle the start action succeeding, while the loop in do_te_invoke is meant to handle a start action failing.

Reply:
> > When the original connection host goes away, every surviving node will detect that via the peer callback. I'm thinking every node should memorize the timestamp at this point (likely a new field in crm_peer_t), since any node could be chosen as the new connection host. This is the time that we'll need to know the remote node was up longer than.
> 
> I assume this is going to have to be a GList or GHashTable, since there could potentially be several remotes that have problems.

If you put it in crm_node_t, the peer cache already handles that. (For daemons other than the controller, lost nodes are automatically removed from the cache, but the controller handles that specially, so I think it would usable for this purpose. If not, we can do a separate GHashTable with a new struct.)

 
> > The DC (possibly newly elected, if the original DC was the host that died) will invoke the scheduler and get the transition graph back. The DC can check for any pseudo-stops with the new meta-attributes, and whether there is a start of the remote connection scheduled in the same transition. If no to either, we can let the purge happen normally.
> 
> I didn't see a great way of checking for a start action in the same transition from down in remote_ra_process_pseudo. So what I did there was add the remote_expired marker and then later on handle actually purging the attributes. The new block in remote_lrm_op_callback is meant to handle the start action succeeding, while the loop in do_te_invoke is meant to handle a start action failing.

Reply:
> At that point it might be better to flip it, and add the meta-attributes to the start action if it exists and there's a pseudo-stop. Then the controller picks up from the start and doesn't need to care about the stop.

And then if we're gonna flip it, I think we also need to then add CRM_meta_expire_attrs="false" to the start action if we found a stop, and then the controller needs to default to purging unless it sees the meta attr (plus the other conditions about uptime).

Reply:
By flip I just meant putting the attribute on the start instead of the stop. The start is when the information actually comes in handy, if we put it on the stop we have to remember it.